crates/cheatcodes/src/inspector.rs

```diff
@@ -217,23 +217,42 @@ pub struct BroadcastableTransaction {
 /// Holds gas metering state.
 #[derive(Clone, Debug, Default)]
 pub struct GasMetering {
+    /// Gas price used in the cheatcode handler to overwrite the gas price separately from the
+    /// gas price in the execution environment.
+    pub gas_price: Option<U256>,
+
     /// True if gas metering is paused.
     pub paused: bool,
-    /// True if gas metering was resumed during the test.
-    pub resumed: bool,
+    /// True if gas metering was resumed or reseted during the test.
+    /// Used to reconcile gas when frame ends (if spent less than refunded).
+    pub touched: bool,
+    /// True if gas metering should be reset to frame limit.
+    pub reset: bool,
     /// Stores frames paused gas.
     pub paused_frames: Vec<Gas>,
+
+    /// Cache of the amount of gas used in previous call.
+    /// This is used by the `lastCallGas` cheatcode.
+    pub last_call_gas: Option<crate::Vm::Gas>,
 }
 
 impl GasMetering {
     /// Resume paused gas metering.
     pub fn resume(&mut self) {
         if self.paused {
             self.paused = false;
-            self.resumed = true;
+            self.touched = true;
         }
         self.paused_frames.clear();
     }
+
+    /// Reset gas to limit.
+    pub fn reset(&mut self) {
+        self.paused = false;
+        self.touched = true;
```

Comment:
does touched need to be set here too? can some combo of pause/resume/reset trigger out of gas without it?

Reply:
yeah, it is basically the same as the pause/resume scenario where we refund after reset, I added a test in https://github.com/foundry-rs/foundry/pull/8750/commits/30311cc8d5ef06efed807e1cc6e07932f5ca717f which would result in `[PASS] testResetNegativeGas() (gas: 18446744073709547520)` if touched not set

Reply:
this though raised an issue with gas reconciliation which was not consistent with gas reset (was setting to remaining gas instead limit), fixed in https://github.com/foundry-rs/foundry/pull/8750/commits/be96028928a5a0e04b513c7597206fdd112b351e gas properly showing 0 now if spent < refunded